compiler/rustc_hir_typeck/src/method/suggest.rs

```diff
@@ -1752,6 +1762,226 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+            }
+        }
+
+        let variants = match item.kind {
+            hir::ItemKind::Struct(_, _hir_generics, variant) => vec![variant],
+            hir::ItemKind::Enum(_, _hir_generics, def) => {
+                def.variants.iter().map(|v| v.data).collect()
+            }
+            _ => return,
+        };
+
+        for variant in variants {
+            for field in variant.fields() {
+                // For every field in the type, we look for those that reference to a type
+                // parameter of the type. For something like `struct Foo<T>(Arc<T>)`, we will
+                // find `Arc<T>`. We check if `Arc` has an implementation for the macro being
+                // derived, and if so if it references `T` in any way. For example, when
+                // deriving `Clone`, `Arc<T>: Clone` doesn't impose `T: Clone`, but when
+                // deriving `PartialEq`, `Arc<T>: PartialEq` does impose `T: PartialEq`. With
+                // this information we can figure out what the appropriate "perfect derive"
+                // bounds could have been, to suggest a manual implementation of the trait.
+
+                let mut my_visitor = TypeParamFinder {
+                    places_found: vec![],
+                    target: self_ty,
+                    tcx: self.tcx,
+                    generics: ty_generics,
+                };
+                my_visitor.visit_ty_unambig(field.ty);
+                if !my_visitor.places_found.is_empty() {
+                    // The current field references the type parameter being bound in the
+                    // original predicate.
+
+                    let mut impls_trait_without_bounds = false;
+                    match field.ty.kind {
+                        hir::TyKind::Path(QPath::Resolved(
+                            _,
+                            hir::Path { res: hir::def::Res::Def(_, ty_def_id), .. },
+                        )) => {
+                            // We have something like `struct Foo<T>(Arc<T>);`. We check if
+                            // `impl<K> Trait for Arc<K>` places a bound on `K`. If it doesn't,
+                            // it means that the perfect derive implementation wouldn't impose
+                            // a bound on `T`, but rather on `Arc:<T>: Trait`.
+                            impls_trait_without_bounds =
+                                self.type_implements_trait_without_bounds(*ty_def_id, d.def_id());
+                        }
+                        _ => {}
+                    }
+                    if let Ok(snippet) = self.tcx.sess.source_map().span_to_snippet(field.ty.span) {
+                        if impls_trait_without_bounds {
+                            bounds
+                                .entry(self_ty.to_string())
+                                .or_default()
+                                .insert("placeholder".to_string());
```

Comment:
"placeholder" is so that we keep the space of types that can implement the trait without obligating the original type parameter to also implement the trait (like for `Arc` and unlike `PartialEq`).